prefect_aws/s3.py

```diff
@@ -337,9 +339,118 @@ def _get_s3_client(self) -> boto3.client:
+            )
         return s3_client
 
+    def _get_bucket_resource(self) -> boto3.resource:
+        if self.minio_credentials:
+            bucket = (
+                self.minio_credentials.get_boto3_session()
+                .resource("s3", endpoint_url=self.endpoint_url)
+                .Bucket(self.bucket_name)
+            )
+
+        elif self.aws_credentials:
+            bucket = (
+                self.aws_credentials.get_boto3_session()
+                .resource("s3")
+                .Bucket(self.bucket_name)
+            )
+        else:
+            raise ValueError(
+                "S3 Bucket requires either a minio_credentials"
+                "field or an aws_credentials field."
+            )
+        return bucket
+
+    async def get_directory(
+        self, from_path: Optional[str] = None, local_path: Optional[str] = None
+    ) -> None:
+        """
+        Copies a folder from the configured S3 bucket to a local directory.
+
+        Defaults to copying the entire contents of the block's basepath to the current
+        working directory.
+
+        Args:
+            from_path: Path in S3 bucket to download from. Defaults to the block's
+                configured basepath.
+            local_path: Local path to download S3 contents to. Defaults to the current
+                working directory.
+        """
+        if from_path is None:
+            from_path = str(self.basepath) if self.basepath else ""
+
+        if local_path is None:
+            local_path = str(Path(".").absolute())
+
+        bucket = self._get_bucket_resource()
+        for object in bucket.objects.filter(Prefix=from_path):
+            if object.key[-1] == "/":
+                # object is a folder and will be created if it contains any objects
+                continue
+            target = os.path.join(
+                local_path,
+                os.path.relpath(object.key, from_path),
+            )
+            if not os.path.exists(os.path.dirname(target)):
+                os.makedirs(os.path.dirname(target))
+            bucket.download_file(object.key, target)
+
+    async def put_directory(
+        self,
+        local_path: Optional[str] = None,
+        to_path: Optional[str] = None,
+        ignore_file: Optional[str] = None,
+    ) -> int:
+        """
+        Uploads a directory from a given local path to the configured S3 bucket in a
+        given folder.
+
+        Defaults to uploading the entire contents the current working directory to the
+        block's basepath.
+
+        Args:
+            local_path: Path to local directory to upload from.
+            to_path: Path in S3 bucket to upload to. Defaults to block's configured
+                basepath.
+            ignore_file: Path to file containing gitignore style expressions for
+                filepaths to ignore.
+
+        """
+        if to_path is None:
+            to_path = str(self.basepath) if self.basepath is not None else ""
```

Comment:
I think it's already a str from `cast_pathlib`

Reply:
Yes, but type-checking doesn't realize that the validator is performing that type casting.